File: 0821_Shortest_Distance_to_a_Character/solution.java
package cn.huang.leetcode;

/*
    821	 Shortest Distance to a Character
 */
public class LC_0821_ShortestDistancetoaCharacter {
    /*
    遍历字符串S,遇到与字符C相等就分别向左/右计算其他字符与该字符的距离，如果其他字符就是C或与目标字符的距离更小的话遍历就终止。
     */
    public int[] shortestToChar(String S, char C) {
        int[] store = new int[S.length()];
        for (int i = 0; i < S.length(); i++) {
            if (S.charAt(i) == C) {
                store[i] = 0;
                left(S, C, store, i);
                right(S, C, store, i);
            }
        }
        return store;
    }

    void left(String S, char C, int[] store, int target) {
        for (int i = target - 1; i >= 0; i--) {
            if (S.charAt(i) == C) break;
            if (store[i] == 0 || store[i] > target - i) store[i] = target - i;
        }
    }

    void right(String S, char C, int[] store, int target) {
        for (int i = target + 1; i < S.length(); i++) {
            if (S.charAt(i) == C) break;
            if (store[i] == 0 || store[i] > i - target) store[i] = i - target;
        }
    }
}

File: 0821_Shortest_Distance_to_a_Character/solution.cpp
/*
821. Shortest Distance to a Character

Level: Easy

https://leetcode.com/problems/shortest-distance-to-a-character
*/
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>

#include <numeric>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

class Solution
{
public:
  /*
    Two Pass
      Time complexity: O(n)
      Space complexity: O(n)
  */
  vector<int> shortestToChar(string S, char C)
  {
    const int n = S.length();
    vector<int> ans(n, INT_MAX);
    int index = -1;
    for (int i = 0; i < n; ++i)
    {
      if (S[i] == C)
        index = i;
      if (index < 0)
        continue;
      ans[i] = abs(i - index);
    }
    index = -1;
    for (int i = n - 1; i >= 0; --i)
    {
      if (S[i] == C)
        index = i;
      if (index < 0)
        continue;
      ans[i] = min(ans[i], abs(i - index));
    }
    return ans;
  }

  vector<int> shortestToChar_v2(string S, char C)
  {
    const int n = S.length();
    vector<int> indices(2 * n); // 0, 1, ..., n - 1, n - 1, n - 2, ..., 0
    std::iota(indices.begin(), indices.begin() + n, 0);
    std::iota(indices.rbegin(), indices.rbegin() + n, 0);
    vector<int> ans(n, INT_MAX);
    int index = -n;
    for (int i : indices)
    {
      if (S[i] == C)
        index = i;
      ans[i] = min(ans[i], abs(i - index));
    }
    return ans;
  }
};